### ISSUE TITLE
srepr not defined for vectors

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#12676]
```
In [1]: from sympy.vector import *

In [2]: e = CoordSysCartesian('e')

In [3]: srepr(e.i)
Out[3]: 'e.i'
```

srepr should give a string that completely recreates the expression (see how it works for other expressions).

**Comment by Hsankesara** [comment]
Hey, I like to do it. Can you please guide me ?

**Comment by Hsankesara** [comment]
Hey @asmeurer, waiting for your reply.Can I go for this?

**Comment by asmeurer** [comment]
Start with https://github.com/sympy/sympy/wiki/Development-workflow, and take a look at the code in sympy/printing/repr.py.

**Comment by ayush1999** [comment]
@asmeurer , srepr shows all operations and operands in the expression(in a tree kind of format). For example:

`from sympy.vector import CoordSysCartesian
from sympy import *
N=CoordSysCartesian('N')
v=3*N.i +N.j
print (srepr(v))`

This gives the output:
`Add(Mul(Integer(3), N.i), N.j)
`
As you can see, the output also contains N,i and N,j.
Can you explain why output 'e.i' in your code should be a mistake?

Thanks!

**Comment by gxyd** [comment]
I think it should be something like (thought not sure):
```python
In [1]:  from sympy.vector import *
In [2]:  e = CoordSysCartesian('e')
In [3]:  srepr(e.i)
Out[3]: 'BaseVector("i", 0, CoordSysCartesian(Symbol("e"), VectorZero(), ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]])), "e_i", "\mathbf{\hat{i}_{e}}")'

In [4]:  BaseVector("i", 0, CoordSysCartesian(Symbol("e"), VectorZero(), ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]])), "e_i", "\\mathbf{\\hat{i}_{e}}")
Out[4]: i
```

It is a large expression but easy to see, look at the `__new__` method of [BaseVector](https://github.com/sympy/sympy/blob/master/sympy/vector/vector.py#L352). As `i` has `type(e.i)` is `sympy.vector.vector.BaseVector`.

**Comment by ayush1999** [comment]
@gxyd, But srepr gives only the operations and operands (in a tree format). why should Out[3] be that (as written in your code) ?

**Comment by gxyd** [comment]
You can consider `BaseVector` to be an operation with operands `("i", 0, CoordSysCartesian(Symbol("e"), VectorZero(), ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]])), "e_i", "\\mathbf{\\hat{i}_{e}}")` in that order, which further has `CoordsSysCartesian` as an operation and its arguments operands. It can also be visualised in a tree-format.

**Comment by asmeurer** [comment]
`srepr` should recreate the expression in a session where only the SymPy names have been imported (for vector it can assume vector names have also been imported). It shouldn't assume that the name `e` has been set to `CoordSysCartesian('e')`. This is generally implemented by printing the full args list. Actually, perhaps the srepr shouldn't even subclass from str printer, but rather have a default printer that just prints the class and the args.

So basically, this should work

```py
from sympy import *
from sympy.vector import *
assert eval(srepr(CoordSysCartesian('e').i)) == CoordSysCartesian('e').i
```

**Comment by gxyd** [comment]
Shouldn't we have `CoordSysCartesian('e').i` be considered an object in itself i.e. I see it odd seeing the `.i` property represented in a tree-structure. Also I have always thought of `srepr` to return something in its tree-structure something that can't be further broken-down. Like in out case `CoordSysCartesian('e')` can be further broken down, like
```python
>>> srepr(CoordSysCartesian('e'))
"CoordSysCartesian(Symbol('e'), VectorZero(), ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]]))"
```
As you can see its further components can't be further broken down like for `Symbol`, `VectorZero`, `ImmutableDenseMatrix(... )`
```python
>>> srepr(Symbol('e'))
"Symbol('e')"
>>> srepr(VectorZero())
'VectorZero()'
>>> srepr(ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]]))
'ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]])'
```
As for the property `eval(srepr( ... )) == ... ` you mention that is still not satisfied.

**Comment by asmeurer** [comment]
The tree structure part is just a side-effect of the way it works. It doesn't have to be that way, it just needs to recreate the object. Although note that if we implement the ideas from https://github.com/sympy/sympy/pull/12524 then attribute access in the output should be avoided.

**Comment by faze-geek** [comment]
Currently on master -
```
>>> e = CoordSys3D('e')
>>> srepr(e.i)
"CoordSys3D(Str('e'), Tuple(ImmutableDenseMatrix([[Integer(1), Integer(0), Integer(0)], [Integer(0), Integer(1), Integer(0)], [Integer(0), Integer(0), Integer(1)]]), VectorZero())).i"
```
I believe this is expected behavior, hence issue maybe closed !?

**Comment by oscarbenjamin** [comment]
Also
```
In [39]: eval(srepr(e.i))
Out[39]: i_e
```
